README.md
# clobank

A Clojure library designed to ... well, that part is up to you.

## Usage

FIXME

## License

Copyright © 2021 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2021-07-27
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2021-07-27
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/clobank/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/clobank/compare/0.1.0...0.1.1


test/clobank/core_test.clj
(ns clobank.core-test
  (:require [clojure.test :refer :all]
            [clobank.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


test/clobank/logic_test.clj
(ns clobank.logic-test
  (:require [clojure.test :refer :all]
            [clobank.logic :refer :all]
            [schema.core :as s]
            [clojure.test.check.generators :as gen]
            [schema-generators.generators :as g]))

(s/set-compile-fn-validation! true)

(defn gerar-amostras-get-total
  ([number]
   (gerar-amostras-get-total number []))
  ([number list]
   (if (> number 0)
     (let [sample (gen/sample  (gen/not-empty (g/generator Transaction)))
           new-list (conj list sample)]
       (gerar-amostras-get-total (dec number) new-list))
     list)))

(deftest test-get-total
  (testing "Uma sequencia vazia"
    (let [map-seq []]
      (is (= 0 (get-total map-seq)))))

  (testing "Uma sequencia vazia"
    (let [map-seq [nil]]
      (is (nil? (get-total map-seq)))))

  (testing "Uma mapa de dados sem o campo AMOUNT"
    (let [map-seq [{:not-amount 1}, {:not-amount 9}, {:not-amount 7}, {:not-amount 3}, {:not-amount 2}, {:not-amount 8}]]
      (is (thrown? java.lang.NullPointerException
                   (get-total map-seq)))))

  (testing "Uma mapa de dados sem alguns contendo o campo AMOUNT"
    (let [map-seq [{:amount 1}, {:amount 9}, {:not-amount 7}, {:not-amount 3}, {:amount 2}, {:amount 8}]]
      (is (thrown? java.lang.NullPointerException
                   (get-total map-seq)))))

  (testing "Uma sequencia de numeros interios positivos"
    (let [map-seq-int [{:amount 1}, {:amount 9}, {:amount 7}, {:amount 3}, {:amount 2}, {:amount 8}]]
      (is (= 30 (get-total map-seq-int)))))

  (testing "Uma sequencia de numeros interios negativos"
    (let [map-seq-int [{:amount -1}, {:amount -9}, {:amount -7}, {:amount -3}, {:amount -2}, {:amount -8}]]
      (is (= -30 (get-total map-seq-int)))))

  (testing "Uma sequencia de numeros decimais positivos"
    (let [map-seq-int [{:amount 0.1}, {:amount 0.9}, {:amount 0.7}, {:amount 0.3}, {:amount 0.2}, {:amount 0.8}]]
      (is (= 3.0 (get-total map-seq-int)))))

  (testing "Uma sequencia de numeros decimais negativos"
    (let [map-seq-int [{:amount -0.1}, {:amount -0.9}, {:amount -0.7}, {:amount -0.3}, {:amount -0.2}, {:amount -0.8}]]
      (is (= -3.0 (get-total map-seq-int)))))

  (testing "Muitas sequencias de numeros aleatorios"
    (doseq [map-seq-int  (gerar-amostras-get-total 100)]
      (is (number? (get-total map-seq-int))))))

(deftest test-transform-group-with-sum
  (testing "Uma mapa vazio"
    (let [data []]
      (is (= {nil 0}
            (transform-group-with-sum data)))))

  (testing "Uma mapa com chave com valores nulos"
    (let [data [:test [nil, nil]]]
      (is (thrown? java.lang.NullPointerException
                   (transform-group-with-sum data)))))

  (testing "Uma mapa com chave com valores nulos"
    (let [data [:test [{:amount 10}, {:amount 100}]]]
      (is (= {:test 110}
             (transform-group-with-sum data)))))
)

(deftest test-group-total-by-category
  (testing "Testando entrada vazia"
    (let [entrada nil]
      (is (= []
           (group-total-by-category entrada)))))

  (testing "Testando entrada valida"
    (let [entrada [{:transactions [
                                   {:amount   100, :category "Educacao"}
                                   {:amount   10,  :category "Alimentacao"}
                                   {:amount   1,   :category "Saude"}]}]]
      (is (= [{"Educacao" 100} {"Alimentacao" 10} {"Saude" 1}]
             (group-total-by-category entrada)))))
)















src/clobank/logic.clj
(ns clobank.logic
  (:require [clobank.db :as cb.db]
            [java-time :as jt]
            [schema.core :as s]))

(s/set-compile-fn-validation! true)

(s/def Transaction
  {:guid     s/Str,
   ;:date     java.time.LocalDateTime,
   :date     s/Str,
   :amount   s/Num,
   :merchant s/Str,
   :category s/Str,
   })

(s/def CreditCard
  {:guid         s/Str,
   :number       s/Str,
   :cvv          s/Int,
   :valid        s/Str,
   :limit        s/Int,
   :transactions [Transaction]})

(s/def CreditCardList  [CreditCard])

(s/def Client
  {:guid         s/Str,
   :name         s/Str,
   :cpf          s/Str,
   :email        s/Str,
   :credit-cards CreditCardList})

(s/def TotalCategory  {s/Keyword s/Num})

(s/def GroupTotalCategory  [TotalCategory])

(defn get-total [value]
  (->> value
       (map :amount)
       (reduce +)))

(s/defn transform-group-with-sum
  [[k, v] :- TotalCategory]
  {k (get-total v)})

(s/defn group-total-by-category :- GroupTotalCategory
  [credit-cards :- CreditCardList]
  (->> credit-cards
       (map :transactions)
       (reduce into [])
       (group-by :category)
       (map transform-group-with-sum)))

(defn get-all-credit-cards
  [customer]
  (->> customer
       :credit-cards
       first))

(defn transactions-by-category [data]
  (->> data
       (map get-all-credit-cards)
       (group-total-by-category)
       println))

(transactions-by-category (cb.db/clobank-database))

(s/defn format-transaction
  [transaction :- Transaction]
  (-> transaction
      (update :date #(str (jt/format "dd/MM/yyyy hh:mm" %)))
      (update :amount #(str "R$ " %))))

(s/defn show-formatted-transactions
  [transaction :- Transaction]
  (println "\nData:" (:date transaction)
           "\nValor:" (:amount transaction)
           "\nEstabelecimento:" (:merchant transaction)
           "\nCategoria:" (:category transaction)))

(s/defn list-transactions
  [data]
  (->> data
       (map get-all-credit-cards)
       (map :transactions)
       (reduce into [])
       (map format-transaction)
       (mapv show-formatted-transactions)))

;(list-transactions (cb.db/clobank-database))
;
;(let [transaction {:guid     "38146102-16c7-4a21-ace8-c17bccecc546",
;                   ;:date     (java-time/local-date-time 2021 07 26 8 34),
;                   :date     "02/12"
;                   :amount   15.00,
;                   :merchant "Bar do Seu Zé",
;                   :category "Alimentação",
;                   }
;      credit-card {:guid         "16f476bc-a4aa-4254-8552-c3b6b99e98d0",
;                     :number       "1111222233334444",
;                     :cvv          123,
;                     :valid        "04/2022",
;                     :limit        5234,
;                     :transactions [transaction]}
;      credit-card-list [credit-card, credit-card]
;      client {:guid         "3fe728fc-216b-402d-9927-d5c2ac0cefa7",
;              :name         "João Clojure",
;              :cpf          "12312312312",
;              :email        "joaoclojure@gmail.com",
;              :credit-cards credit-card-list}]
;
;  (println "\nTeste schema transaction:" (s/validate Transaction transaction))
;  (println "Teste schema Credit Card:" (s/validate CreditCard credit-card))
;  (println "Teste schema Credit Card List:" (s/validate CreditCardList credit-card-list))
;  (println "Teste schema Client:" (s/validate Client client)))


src/clobank/alternative-logic.clj
(ns clobank.alternative
  (:require [clobank.db :as cb.db]
            [java-time :as jt]))

(defn get-total [v]
  (->> v
       (map #(get % :amount))
       (reduce +)))

(defn transform-group-with-sum
  [[k, v]]
  {
   k (get-total v)})

(defn group-total-by-category
  [ccs]
  (->> ccs
       (map :transactions)
       (reduce into [])
       (group-by :category)
       (map transform-group-with-sum)))

(defn group-total-by-merchant
  [ccs]
  (->> ccs
       (map :transactions)
       (reduce into [])
       (group-by :merchant)))

(defn monthly-charge
  [css]
  (->> css
       (map :transactions)
       (reduce into [])
       (when (= (jt/local-date :month-of-year)))))

(defn format-transaction
  [transaction]
  (-> transaction
      (update :date #(str (jt/format "dd/MM/yyyy hh:mm" %)))
      (update :amount #(str "R$ " %))))

(defn show-formatted-transactions
  [transaction]
  (println transaction)
  (println "\nData:" (:date transaction)
           "\nValor:" (:amount transaction)
           "\nEstabelecimento:" (:merchant transaction)
           "\nCategoria:" (:category transaction)))

(defn get-all-credit-cards
  [customer]
  (->> customer
       :credit-cards
       first))

(defn transactions-by-category []
  (->> (cb.db/clobank-database)
       (map get-all-credit-cards)
       (group-total-by-category)
       println))

(defn transactions-by-merchant []
  (->> (cb.db/clobank-database)
       (map get-all-credit-cards)
       (group-total-by-merchant)
       println))

(defn list-transactions []
  (->> (cb.db/clobank-database)
       (map get-all-credit-cards)
       (map :transactions)
       (reduce into [])
       (map format-transaction)
       println
       ;(map show-formatted-transactions)
       ))


src/clobank/db.clj
(ns clobank.db)
(require '[java-time :as java-time])
(def customer1 {:guid         "3fe728fc-216b-402d-9927-d5c2ac0cefa7",
                :name         "João Clojure",
                :cpf          "12312312312",
                :email        "joaoclojure@gmail.com",
                :credit-cards [{:guid         "16f476bc-a4aa-4254-8552-c3b6b99e98d0",
                                :number       "1111222233334444",
                                :cvv          123,
                                :valid        "04/2022",
                                :limit        5234,
                                :transactions [{:guid     "38146102-16c7-4a21-ace8-c17bccecc546",
                                                :date     (java-time/local-date-time 2021 07 26 8 34),
                                                :amount   15.00,
                                                :merchant "Bar do Seu Zé",
                                                :category "Alimentação",
                                                },
                                               {:guid     "1f4bab0c-b0c4-4723-ba3f-be9fe7a7acb2",
                                                :date     (java-time/local-date-time 2021 07 26 12 0),
                                                :amount   320.40,
                                                :merchant "Clínica Sorriso Feliz",
                                                :category "Saúde",
                                                },
                                               {:guid     "f6e8b759-3059-43a0-aee4-96916de989d6",
                                                :date     (java-time/local-date-time 2021 07 27 9 48),
                                                :amount   110.20,
                                                :merchant "Hi Lorena Idiomas",
                                                :category "Educação",
                                                },
                                               {:guid     "cfff8b7c-9186-4bc3-83a4-870eb94040dc",
                                                :date     (java-time/local-date-time 2021 06 28 15 22),
                                                :amount   70.00,
                                                :merchant "Mosca Frita",
                                                :category "Alimentação",
                                                }]}]})

(def customer2 {:guid         "asdasd31-216b-402d-9927-d5c2ac0cefa2",
                :name         "Maria Clojure",
                :cpf          "12312312312",
                :email        "ma_clojure@gmail.com",
                :credit-cards [{:guid         "16f476bc-a4aa-4254-8552-c3b6b99e98d1",
                                :number       "1111222233334444",
                                :cvv          123,
                                :valid        "04/2022",
                                :limit        5234,
                                :transactions [{:guid     "38146102-16c7-4a21-ace8-c17bccecc516",
                                                :date     (java-time/local-date-time 2021 07 26 8 34),
                                                :amount   25.00,
                                                :merchant "Bar do Seu Zé",
                                                :category "Alimentação",
                                                },
                                               {:guid     "1f4bab0c-b0c4-4723-ba3f-be9fe7a7a3b2",
                                                :date     (java-time/local-date-time 2021 06 26 12 0),
                                                :amount   420.40,
                                                :merchant "Clínica Sorriso Feliz",
                                                :category "Saúde",
                                                },
                                               {:guid     "f6e8b759-3059-43a0-aee4-96916d3989d6",
                                                :date     (java-time/local-date-time 2021 07 27 9 48),
                                                :amount   210.20,
                                                :merchant "Hi Lorena Idiomas",
                                                :category "Educação",
                                                },
                                               {:guid     "cfff8b74-9186-4bc3-83a4-870eb94040dc",
                                                :date     (java-time/local-date-time 2021 07 28 15 22),
                                                :amount   80.00,
                                                :merchant "Mosca Frita",
                                                :category "Alimentação",
                                                }]}]})

(defn clobank-database []
  [customer1, customer2])